launch/shell_unix.go

```diff
@@ -0,0 +1,71 @@
+// +build linux darwin
+
+package launch
+
+import (
+	"fmt"
+
+	"github.com/pkg/errors"
+
+	"github.com/buildpacks/lifecycle/api"
+)
+
+var (
+	bashCommandWithScript = `exec bash -c "$@"` // for processes w/o argsument
+	profileGlob           = "*"
+	appProfile            = ".profile"
+)
+
+func (l *Launcher) execWithShell(self string, process Process, profiles []string) error {
+	launcher := ""
+	for _, profile := range profiles {
+		launcher += fmt.Sprintf("source \"%s\"\n", profile)
+	}
+	bashCommand, err := l.bashCommand(process)
+	if err != nil {
+		return errors.Wrap(err, "determine bash command")
+	}
+	launcher += bashCommand
+	if err := l.Exec("/bin/bash", append([]string{
+		"bash", "-c",
+		launcher, self, process.Command,
+	}, process.Args...), l.Env.List()); err != nil {
+		return errors.Wrap(err, "bash exec")
+	}
+	return nil
+}
+
+func (l *Launcher) bashCommand(process Process) (string, error) {
+	if len(process.Args) == 0 {
+		return bashCommandWithScript, nil
+	}
+	if process.BuildpackID == "" {
+		return bashCommandWithTokens(len(process.Args) + 1), nil
+	}
+	for _, bp := range l.Buildpacks {
+		if bp.ID != process.BuildpackID {
+			continue
+		}
+		bpAPI, err := api.NewVersion(bp.API)
+		if err != nil {
+			return "", fmt.Errorf("failed to parse api '%s' of buildpack '%s'", bp.API, bp.ID)
+		}
+		if isLegacyProcess(bpAPI) {
+			return bashCommandWithScript, nil
+		}
+		return bashCommandWithTokens(len(process.Args) + 1), nil
+	}
+	return "", fmt.Errorf("process type '%s' provided by unknown buildpack '%s'", process.Type, process.BuildpackID)
+}
+
+func isLegacyProcess(bpAPI *api.Version) bool {
```

Comment:
I notice we don't do this check in shell_windows.go, is it because Windows support is so new? While I doubt we'd get buildpacks supporting an older API, it might be worthwhile to have this check, if it's considered supported.

Reply:
Windows buildpack support was introduced in buildpack API `0.4`. Adding an analogous legacy execution to windows would require some complicated additional work and it probably isn't worth doing (considering we want to move to the new API). We may want to add a check to ensuring that windows buildpacks never have API < 0.4. But, we probably want do that a build time so folks aren't surprised at launch, and therefore I think we should punt it to a separate PR.